source/common/json/json_internal.cc

```diff
@@ -0,0 +1,668 @@
+
+  bool isType(Type type) const { return type == type_; }
+  void checkType(Type type) const {
+    if (!isType(type)) {
+      throw Exception(fmt::format(
+          "JSON field from line {} accessed with type '{}' does not match actual type '{}'.",
+          line_number_start_, typeAsString(type), typeAsString(type_)));
+    }
+  }
+
+  // Value return type functions.
+  std::string stringValue() const {
+    checkType(Type::String);
+    return value_.string_value_;
+  }
+  std::vector<FieldSharedPtr> arrayValue() const {
+    checkType(Type::Array);
+    return value_.array_value_;
+  }
+  bool booleanValue() const {
+    checkType(Type::Boolean);
+    return value_.boolean_value_;
+  }
+  double doubleValue() const {
+    checkType(Type::Double);
+    return value_.double_value_;
+  }
+  int64_t integerValue() const {
+    checkType(Type::Integer);
+    return value_.integer_value_;
+  }
+
+  nlohmann::json asJsonDocument() const;
+  static void buildJsonDocument(const Field& field, nlohmann::json& value);
+
+  uint64_t line_number_start_ = 0;
+  uint64_t line_number_end_ = 0;
+  const Type type_;
+  Value value_;
+};
+
+/**
+ * Consume events from SAX callbacks to build JSON Field.
+ */
+class ObjectHandler : public nlohmann::json_sax<nlohmann::json> {
+public:
+  ObjectHandler() = default;
+
+  bool start_object(std::size_t) override;
+  bool end_object() override;
+  bool key(std::string& val) override;
+  bool start_array(std::size_t) override;
+  bool end_array() override;
+  bool boolean(bool value) override { return handleValueEvent(Field::createValue(value)); }
+  bool number_integer(int64_t value) override {
+    return handleValueEvent(Field::createValue(static_cast<int64_t>(value)));
+  }
+  bool number_unsigned(uint64_t value) override {
+    if (value > static_cast<uint64_t>(std::numeric_limits<int64_t>::max())) {
+      throw Exception(fmt::format("JSON value from line {} is larger than int64_t (not supported)",
+                                  line_number_));
+    }
+    return handleValueEvent(Field::createValue(static_cast<int64_t>(value)));
+  }
+  bool number_float(double value, const std::string&) override {
+    return handleValueEvent(Field::createValue(value));
+  }
+  bool null() override { return handleValueEvent(Field::createNull()); }
+  bool string(std::string& value) override { return handleValueEvent(Field::createValue(value)); }
+  bool binary(binary_t&) override { return false; }
+  bool parse_error(std::size_t, const std::string& token,
+                   const nlohmann::detail::exception& ex) override {
+    // Errors are formatted like "[json.exception.parse_error.101] parse error at line x, column y:
+    // error string."
+    absl::string_view error = ex.what();
+    // Extract position information (line, column, token) in the error.
+    auto start = error.find("line");
+    auto npos = error.find(":");
+    error_position_ = absl::StrCat(error.substr(start, npos - start), ", token ", token);
```

Comment:
would like to see some checking of these `find()` results against `std::npos` or `absl::npos`. And while we're at it, let's not also have a local called `npos` :)

Reply:
done, i added a reference, there should always be a ": error string" but I found in the documentation that position information may not be present, so that's handled now.

Thank you!